core/src/main/java/com/linecorp/armeria/client/routing/ZookeeperEndpointGroup.java

```diff
@@ -0,0 +1,154 @@
+ * under the License.
+ */
+package com.linecorp.armeria.client.routing;
+
+import static java.util.Objects.requireNonNull;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Function;
+
+import org.apache.zookeeper.AsyncCallback.StatCallback;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.linecorp.armeria.client.Endpoint;
+
+/**
+ *  a zookeeper {@link EndpointGroup} implementation .
+ *  it will get the {@link EndpointGroup} information from a zookeeper zNode, any update to it will
+ *  reflect asynchronously to the system
+ */
+public class ZookeeperEndpointGroup implements EndpointGroup, Watcher, StatCallback {
+
+    private static final Logger logger = LoggerFactory.getLogger(ZookeeperEndpointGroup.class);
+
+    private String hostPort;
+    private String znode;
+    private List<Endpoint> endpointList;
+    private ZooKeeper zooKeeper;
+    byte[] prevData;
+    Function<byte[], List<Endpoint>> nodeValueToEndpointList;
+
+    final CountDownLatch connectionLatch = new CountDownLatch(1);
+
+    /**
+     * create a zookeeper endpoint group
+     * @param hostPort zookeeper hostPort eg. localhost:2181
+     * @param zNode a zNode path eg. /groups/productionGroups
+     * @param sessionTimeout   zookeeper session timeout
+     * @param nodeValueToEndpointList a function to convert zNode value to a List of {@link Endpoint}
+     */
+    public ZookeeperEndpointGroup(String hostPort, String zNode, int sessionTimeout,
+                                  Function<byte[], List<Endpoint>> nodeValueToEndpointList) {
+
+        requireNonNull(hostPort);
+        requireNonNull(zNode);
+        requireNonNull(nodeValueToEndpointList);
+        this.hostPort = hostPort;
+        this.znode = zNode;
+        this.nodeValueToEndpointList = nodeValueToEndpointList;
+        try {
+            zooKeeper = new ZooKeeper(hostPort, sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    connectionLatch.countDown();
+                }
+
+            });
+            connectionLatch.await();
+
+            Stat stat = zooKeeper.exists(zNode, this);
+            byte[] nodeData;
+            if (stat != null) {
+                nodeData = zooKeeper.getData(zNode, false, null);
+                endpointList = nodeValueToEndpointList.apply(nodeData);
+            }
+
+        } catch (IOException e) {
+            logger.error("error occurred while connecting zooKeeper " + e.getMessage());
+            e.printStackTrace();
```

Comment:
Instead of logging, please throw a new `EndpointGroupException`. For example:

```java
throw new EndpointGroupException("failed to connect to ZooKeeper: " + hostPort, e);
```